### PR TITLE
Add new event code 

### PR DESCRIPTION
**Pull request by puranban**
## Addresses 
- https://github.com/idmc-labs/helix2.0-meta/issues/963
- https://github.com/idmc-labs/helix2.0-meta/issues/967
- https://github.com/idmc-labs/helix2.0-meta/issues/971
- https://github.com/idmc-labs/helix2.0-meta/issues/972
- https://github.com/idmc-labs/helix2.0-meta/issues/989
- https://github.com/idmc-labs/helix2.0-meta/issues/996

## Depends on

- https://github.com/idmc-labs/helix-server/pull/543

## Changes

* Remove old event code section.
* Add new inputs section for event code.
* Add new file for `EventCodeInput` component.
* Add validations for event code country.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] permission checks
- [x] translations

### CODE REVIEW

File: src/components/forms/EventForm/index.tsx [hunk 1]
```diff
@@ -352,6 +372,17 @@ const schema: FormSchema = {
             osvSubType: [nullCondition],
             actor: [nullCondition],
             otherSubType: [nullCondition],
+            eventCodes: {
+                keySelector: (c: PartialForm<EventCode>) => c.clientId as string,
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:376]: We don't need to pass the typing here explicitly

File: src/components/forms/EventForm/index.tsx [hunk 2]
```diff
@@ -352,6 +372,17 @@ const schema: FormSchema = {
...
+                member: (): EventCodesSchemaMember => ({
+                    fields: (): EventCodeSchemaFields => ({
+                        clientId: [],
+                        country: [requiredCondition],
+                        eventCodeType: [requiredCondition],
+                        eventCode: [requiredCondition],
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:382]:
```suggestion
                        eventCode: [requiredStringCondition],
```

File: src/components/forms/EventForm/index.tsx [hunk 3]
```diff
@@ -325,12 +328,29 @@ const conflict: CrisisType = 'CONFLICT';
...
+    country: string;
+    eventCodeType: string;
+    eventCode: string;
+}
+
 type EventFormFields = CreateEventMutationVariables['event'];
-type FormType = PurgeNull<PartialForm<WithId<EventFormFields>>>;
+interface EventFormFieldsWithEventCode extends EventFormFields {
+    eventCodes: EventCode[];
+}
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:341]: Let's add a TODO that we need to remove this once it's implemented on the server.

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                    onChange={onValueChange}
+                    onOptionsChange={setCountryOptions}
+                    error={undefined}
+                    disabled={disabled}
+                    readOnly={readOnly}
+                />
+                <SelectInput
+                    label="Type *"
+                    name=""
+                    options={[]}
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:91]: Let's pass default values here and add a NOTE to get this from server.
NOTE: We should define the options on the parent and pass it to this component.

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                    value={undefined}
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:92]: The `value` prop should be read from the `value.type`.
Similar comment for other props that are not defined.

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                />
+            </Row>
+            <Button
+                name={index}
+                onClick={onRemove}
+                compact
+                transparent
+            >
+                <IoTrash />
+            </Button>
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:117]: Let's pass disabled and readOnly props

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                <SelectInput
+                    label="Country *"
+                    name="country"
+                    options={options}
+                    value={value.country}
+                    keySelector={keySelector}
+                    labelSelector={labelSelector}
+                    onChange={onValueChange}
+                    onOptionsChange={setCountryOptions}
+                    error={undefined}
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:84]: Let's pass `error` fields too.

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        ),
+    ).filter(isDefined);
+
+    return (
+        <Row
+            singleColumnNoGrow
+            className={_cs(className, styles.eventCode)}
+        >
+            <Row className={styles.input}>
+                <SelectInput
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:75]: We should also add a NonFieldError component

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            (country) => country.id === countryId,
+        ),
+    ).filter(isDefined);
+
+    return (
+        <Row
+            singleColumnNoGrow
+            className={_cs(className, styles.eventCode)}
+        >
+            <Row className={styles.input}>
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:74]: Why are we nesting Row inside a Row?

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        error,
+        disabled,
+        readOnly,
+    } = props;
+    const onValueChange = useFormObject(index, onChange, defaultValue);
+    const options = countryIds?.map(
+        (countryId) => countryOptions?.find(
+            (country) => country.id === countryId,
+        ),
+    ).filter(isDefined);
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:67]: The complexity is `O(N^2)` and it's not entirely clear what we are want to achieve.
Why not just use a filter instead?

Comment (tnagorra) [reply src/components/forms/EventForm/EventCodeInput/index.tsx:67]: Also, this logic should be implemented in the parent component

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/forms/EventForm/EventCodeInput/index.tsx (hunk 9 above)
Comment (puranban) [reply src/components/forms/EventForm/EventCodeInput/index.tsx:74]: To make consistent styling of remove icon, same as other `eventCodeInput`.

File: src/components/forms/EventForm/index.tsx [hunk 11]
```diff
@@ -899,26 +986,56 @@ function EventForm(props: EventFormProps) {
...
+                    >
+                        Add Event Code
+                    </Button>
+                )}
+            >
+                {eventCodeLimitExceed && (
+                    <NonFieldWarning>
+                        Cannot add more than 50 event codes
+                    </NonFieldWarning>
+                )}
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:1018]: The error should come directly from form.
We should use NonFieldError to show this error.

File: src/components/forms/EventForm/index.tsx [hunk 12]
```diff
@@ -764,6 +815,42 @@ function EventForm(props: EventFormProps) {
...
+                'eventCodes' as const,
+            );
+        }, [onValueChange],
+    );
+
+    const handleCountryChange = useCallback(
+        (countryId) => {
+            const filterEventCodeByCountry = value.eventCodes?.filter(
+                (c) => countryId.includes(c.country),
+            );
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:842]: Let's read old values from the function argument on onValueChange

File: src/components/forms/EventForm/index.tsx [hunk 13]
```diff
@@ -764,6 +815,42 @@ function EventForm(props: EventFormProps) {
...
+                (oldValue: PartialForm<EventCode[]> | undefined) => (
+                    [...(oldValue ?? []), newEventCodeItem]
+                ),
+                'eventCodes' as const,
+            );
+        }, [onValueChange],
+    );
+
+    const handleCountryChange = useCallback(
+        (countryId) => {
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:839]:
```suggestion
        (countryIds) => {
```

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    useFormObject,
+} from '@togglecorp/toggle-form';
+
+import Row from '#components/Row';
+import NonFieldError from '#components/NonFieldError';
+import { CountryOption } from '#components/selections/CountrySelectInput';
+
+import styles from './styles.css';
+
+type EventCode = {
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:18]: Let's add a TODO to remove this later

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+type EventCodeSchema = PartialForm<EventCode>;
+
+const keySelector = (d: CountryOption) => d.id;
+const labelSelector = (d: CountryOption) => d.idmcShortName;
+
+const defaultValue: EventCodeSchema = {
+    uuid: randomString(),
+};
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:32]: We cannot use randomString here as it will be the same for all of the values.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+// TODO: remove once it's implemented on the server.
+type EventCode = {
+    uuid: string;
+    country: string;
+    eventCodeType: string;
+    eventCode: string;
+}
+
+type EventCodeSchema = PartialForm<EventCode>;
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:26]:
```suggestion
type PartialEventCode = PartialForm<EventCode>;
```

File: src/components/forms/EventForm/index.tsx [hunk 17]
```diff
@@ -764,12 +820,51 @@ function EventForm(props: EventFormProps) {
...
+                (oldValue: PartialForm<EventCode[]> | undefined) => (
+                    [...(oldValue ?? []), newEventCodeItem]
+                ),
+                'eventCodes' as const,
+            );
+        }, [onValueChange],
+    );
+
+    const handleCountryChange = useCallback(
+        (countryIds) => {
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:844]: Let's add type for countryIds

File: src/components/forms/EventForm/index.tsx [hunk 18]
```diff
@@ -899,26 +994,55 @@ function EventForm(props: EventFormProps) {
...
             />
+            <Section
+                subSection
+                heading="Event Code"
+                actions={(
+                    <Button
+                        name="addEventCode"
+                        onClick={handleEventCodeAddButtonClick}
+                        disabled={isNotDefined(value.countries)
+                        || (value?.eventCodes?.length ?? 0) > 50}
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:1016]: Let's also disable this when form is disabled or readonly

File: src/components/forms/EventForm/index.tsx [hunk 19]
```diff
@@ -764,12 +820,51 @@ function EventForm(props: EventFormProps) {
...
+
     const children = (
         <>
             {loading && <Loading absolute />}
             <NonFieldError>
                 {error?.$internal}
             </NonFieldError>
+            <NonFieldError>
+                {error?.fields?.eventCodes?.$internal}
+            </NonFieldError>
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:867]: This is repeated twice. Let's remove this one.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/forms/EventForm/index.tsx (hunk 19 above)
Comment (puranban) [reply src/components/forms/EventForm/index.tsx:868]: done

File: src/components/forms/EventForm/EventCodeInput/index.tsx (hunk 14 above)
Comment (puranban) [reply src/components/forms/EventForm/EventCodeInput/index.tsx:19]: done

File: src/components/forms/EventForm/EventCodeInput/index.tsx (hunk 8 above)
Comment (puranban) [reply src/components/forms/EventForm/EventCodeInput/index.tsx:88]: done

File: src/components/forms/EventForm/index.tsx (hunk 19 above)
Comment (tnagorra) [reply src/components/forms/EventForm/index.tsx:868]: Please delete this one and undo deleting the other block

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+        readOnly,
+    } = props;
+
+    const defaultValue = useMemo(() => ({ uuid: randomString() }), []);
+    const onValueChange = useFormObject(index, onChange, defaultValue);
+
+    const options = useMemo(
+        () => countryOptions?.filter((country) => countryIds?.includes(country.id)),
+        [countryIds, countryOptions],
+    );
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:77]: Let's move this outside this component. We will have to compute this on each item un-necessarily.

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+            <Row className={styles.input}>
+                <SelectInput
+                    label="Country *"
+                    name="country"
+                    options={options}
+                    value={value.country}
+                    keySelector={keySelector}
+                    labelSelector={labelSelector}
+                    onChange={onValueChange}
+                    onOptionsChange={setCountryOptions}
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:96]: As this is a SelectInput, we don't need to pass onOptionsChange

File: src/components/forms/EventForm/EventCodeInput/index.tsx [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+            <Button
+                name={index}
+                onClick={onRemove}
+                compact
+                transparent
+                disabled={disabled}
+                readOnly={readOnly}
+            >
+                <IoTrash />
+            </Button>
```
Comment (tnagorra) [inline src/components/forms/EventForm/EventCodeInput/index.tsx:134]: Let's add "title" on icon buttons

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/forms/EventForm/index.tsx (hunk 19 above)
Comment (tnagorra) [reply src/components/forms/EventForm/index.tsx:872]: @puranban Let's move this block inside the Event Code section.

File: src/components/forms/EventForm/index.tsx [hunk 23]
```diff
@@ -899,26 +999,52 @@ function EventForm(props: EventFormProps) {
...
             />
+            <Section
+                subSection
+                heading="Event Code"
+                actions={(
+                    <Button
+                        name="addEventCode"
+                        onClick={handleEventCodeAddButtonClick}
+                        disabled={isNotDefined(value.countries)
+                        || (value?.eventCodes?.length ?? 0) > 50
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:1021]: Let's define a MAX_EVENT_CODES variable and also use this on the schema.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/forms/EventForm/index.tsx [hunk 24]
```diff
@@ -352,6 +375,23 @@ const schema: FormSchema = {
...
+                    fields: (): EventCodeSchemaField => ({
+                        uuid: [],
+                        country: [requiredCondition],
+                        eventCodeType: [requiredCondition],
+                        eventCode: [requiredStringCondition],
+                    }),
+                }),
+                validation: (val) => {
+                    if (isDefined(val) && val.length > MAX_EVENT_CODES) {
+                        return 'Cannot add more than 50 event codes.';
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:390]:
```suggestion
                        return `Cannot add more than ${MAX_EVENT_CODES} event codes.`;
```

File: src/components/forms/EventForm/index.tsx [hunk 25]
```diff
@@ -764,6 +821,47 @@ function EventForm(props: EventFormProps) {
...
+                    [...(oldValue ?? []), newEventCodeItem]
+                ),
+                'eventCodes' as const,
+            );
+        }, [onValueChange],
+    );
+
+    const handleCountryChange = useCallback(
+        (countryIds: PartialForm<FormType['countries']>) => {
+            onValueChange(() => ([...(countryIds ?? [])]), 'countries' as const);
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:846]:
```suggestion
            onValueChange(countryIds ?? [], 'countries' as const);
```

File: src/components/forms/EventForm/index.tsx [hunk 26]
```diff
@@ -764,6 +821,47 @@ function EventForm(props: EventFormProps) {
...
+            onValueChange(
+                (oldValue: PartialForm<EventCode[]>| undefined) => {
+                    const filterEventCodeByCountry = oldValue?.filter(
+                        // eslint-disable-next-line
```
Comment (tnagorra) [inline src/components/forms/EventForm/index.tsx:850]: Why are we disabling eslint? We should only disable specific rule

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Review by tnagorra — APPROVED** [review]

**Review by tnagorra — APPROVED** [review]

**Pull request merged**